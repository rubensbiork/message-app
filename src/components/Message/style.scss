.message-wrapper {
  margin-top: 30px;
  margin-bottom: 30px;
  position: relative;

  &:hover .destroy,
  &:hover .rename {
    display: block;
  }

  .edit {
  	display: none;
  }

  &.editing {
  	border-bottom: none;
  	padding: 0;
  }

  &.editing .edit {
    position: relative;
    display: block;
    width: calc(100% - 40px) !important;
    top: -10px;
  }

  &.editing .view {
  	display: none;
  }

  .img-circle {
    position: relative;
    vertical-align: top;
    height: 45px;
    width: 45px;
    border-radius: 50%;
    box-shadow: 3px 3px 10px #979b9e;
    margin-left: 21px;
  }

  .edit {
  	width: calc(100% - 90px) !important;
  	height: 45px;
  	border: 1px solid #e6e6e6;
  	background-color: #eceff1;
  	border-radius: 10px;
  	font-size: 20px;
  	padding-left: 20px
  }

  .speech-bubble {
    position: relative;
    background: #cccece;
    border-radius: 8px;
    min-height: 47px;
    width: calc(100% - 100px);
    margin-left: 22px;
    display: inline-block;
    top: -10px;

    .main-message {
      font-size: 16px;
      overflow-wrap: break-word;
      padding: 15px;
    }
  }

  .speech-bubble:after {
    content: '';
    position: absolute;
    left: 0;
    top: 35px;
    width: 13px;
    height: 0;
    border: 16px solid transparent;
    border-right-color: #cccece;
    border-left: 0;
    border-bottom: 0;
    margin-top: -17px;
    margin-left: -29px;
  }

  .time-message {
    position: relative;
    float: right;
    top: -10px;
    right: 17px;
    font-size: 13px;
  }

  .rename {
    display: none;
    position: absolute;
    cursor: pointer;
    stroke: #635d5d;
    top: 4px;
    right: 68px;
    transition: color 0.2s ease-out;

    &:hover {
      stroke: #758c8c;
    }

    &:after {
      display: block;
    }

    .edit-icon {
      width: 20px;
      height: 20px;
    }
  }

  .destroy {
    display: none;
    position: absolute;
    cursor: pointer;
    color: #635d5d;
    top: -7px;
    right: 26px;
    font-size: 30px;
    transition: color 0.2s ease-out;

    &:hover {
      color: #ff0000;
    }

    &:after {
      content: 'Ã—';
    }
  }
}
