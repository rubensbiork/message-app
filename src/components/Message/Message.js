import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import {
  createFragmentContainer,
  graphql,
} from 'react-relay';

import RemoveMessageMutation from '../../mutations/RemoveMessageMutation';
import RenameMessageMutation from '../../mutations/RenameMessageMutation';

import MessageInput from '../MessageInput';
import Icon from 'svg-react-loader?name=Icon!../../assets/edit.svg';
import utcToLocaTime from '../../utils/dateTime';

import './style.scss';

export class Message extends PureComponent {
  state = {
    isEditing: false,
  };
  _onDestroyClick = () => {
    this._removeMessage();
  };
  _onEditClick = () => {
    this._setEditMode(true);
  };
  _onInputCancel = () => {
    this._setEditMode(false);
  };
  _onInputSave = (text) => {
    this._setEditMode(false);
    RenameMessageMutation.commit(
      this.props.relay.environment,
      text,
      this.props.message,
    );
  };
  _removeMessage() {
    RemoveMessageMutation.commit(
      this.props.relay.environment,
      this.props.message,
      this.props.viewer,
    );
  }
  _setEditMode = (isEditing) => {
    this.setState({isEditing});
  };

  renderInput() {
    return (
      <MessageInput
        className="edit"
        existingValue={this.props.message.text}
        onCancel={this._onInputCancel}
        onSave={this._onInputSave}
      />
    );
  }

  renderBallon() {
    const {
      avatar,
    } = this.props.viewer;
    const {
      text,
      createdAt,
    } = this.props.message;
    return (
      <div>
        <img className="img-circle" src={avatar}/>
        <div className="speech-bubble">
          <div className="main-message">{text}</div>
        </div>
        <div className="time-message">{utcToLocaTime(createdAt)}</div>
        <div
          className="rename"
          onClick={this._onEditClick}
        >
          <Icon className="edit-icon" />
        </div>
        <button
          className="destroy"
          onClick={this._onDestroyClick}
        />
      </div>
    );
  }

  render() {
    const {
      isEditing,
    } = this.state;

    return (
      <li
        className={`message-wrapper ${isEditing ? 'editing' : ''}`}
      >
        <div>
          {isEditing ? this.renderInput() : this.renderBallon()}
        </div>
      </li>
    );
  }
}

Message.defaultProps = {
  message: {},
  viewer: {},
};

Message.propTypes = {
  message: PropTypes.shape({}),
  viewer: PropTypes.shape({}),
};

export default createFragmentContainer(Message, {
  message: graphql`
    fragment Message_message on Message {
      id,
      text,
      createdAt,
    }
  `,
  viewer: graphql`
    fragment Message_viewer on User {
      id,
      totalCount,
      avatar,
    }
  `,
});
