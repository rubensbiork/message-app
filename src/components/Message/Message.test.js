import React from 'react';
import { shallow } from 'enzyme';
import { Message } from './Message';

const wrapper = props => shallow(<Message {...props} />);

const data = {
  message: {
    text: 'some-text',
    createdAt: '2017-07-22T18:05:26Z',
  },
  viewer: {
    avatar: 'some-url',
  },
};

describe('components/<Message />', () => {
  it('Should render', () => {
    const component = wrapper({...data});

    expect(
      component,
    ).toMatchSnapshot();
  });

  it('Should set the avatar from given URL', () => {
    const component = wrapper({...data});

    expect(
      component.find('.img-circle').props().src
    ).toBe('some-url');
  });

  it('Should display localized and formatted time', () => {
    const component = wrapper({...data});

    expect(
      component.find('.time-message').text()
    ).toBe('22-7-2017 20:05');
  });

  it('Should display input after clicking in edit icon', () => {
    const component = wrapper({...data});

    const editIcon = component.find('.rename');
    editIcon.simulate('click');

    expect(
      component.state().isEditing
    ).toBeTruthy();

    expect(
      component
    ).toMatchSnapshot();
  });
});
