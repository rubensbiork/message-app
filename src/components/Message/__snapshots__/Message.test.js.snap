// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/<Message /> Should display input after clicking in edit icon 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="message-wrapper editing"
>
    <div>
        <MessageInput
            className="edit"
            existingValue="some-text"
            onCancel={[Function]}
            onDelete={[Function]}
            onSave={[Function]}
            placeholder=""
        />
    </div>
</li>,
  "nodes": Array [
    <li
      className="message-wrapper editing"
>
      <div>
            <MessageInput
                  className="edit"
                  existingValue="some-text"
                  onCancel={[Function]}
                  onDelete={[Function]}
                  onSave={[Function]}
                  placeholder=""
            />
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Message
        message={
                Object {
                        "createdAt": "2017-07-22T18:05:26Z",
                        "text": "some-text",
                      }
        }
        viewer={
                Object {
                        "avatar": "some-url",
                      }
        }
/>,
      "_debugID": 7,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Message {
        "_onDestroyClick": [Function],
        "_onEditClick": [Function],
        "_onInputCancel": [Function],
        "_onInputSave": [Function],
        "_reactInternalInstance": [Circular],
        "_setEditMode": [Function],
        "context": Object {},
        "props": Object {
          "message": Object {
            "createdAt": "2017-07-22T18:05:26Z",
            "text": "some-text",
          },
          "viewer": Object {
            "avatar": "some-url",
          },
        },
        "refs": Object {},
        "state": Object {
          "isEditing": true,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 4,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li
          className="message-wrapper editing"
>
          <div>
                    <MessageInput
                              className="edit"
                              existingValue="some-text"
                              onCancel={[Function]}
                              onDelete={[Function]}
                              onSave={[Function]}
                              placeholder=""
                    />
          </div>
</li>,
        "_debugID": 8,
        "_renderedOutput": <li
          className="message-wrapper editing"
>
          <div>
                    <MessageInput
                              className="edit"
                              existingValue="some-text"
                              onCancel={[Function]}
                              onDelete={[Function]}
                              onSave={[Function]}
                              placeholder=""
                    />
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Message
    message={
        Object {
            "createdAt": "2017-07-22T18:05:26Z",
            "text": "some-text",
          }
    }
    viewer={
        Object {
            "avatar": "some-url",
          }
    }
/>,
}
`;

exports[`components/<Message /> Should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    className="message-wrapper "
>
    <div>
        <div>
            <img
                className="img-circle"
                src="some-url"
            />
            <div
                className="speech-bubble"
            >
                <div
                    className="main-message"
                >
                    some-text
                </div>
            </div>
            <div
                className="time-message"
            >
                22-7-2017 20:05
            </div>
            <div
                className="rename"
                onClick={[Function]}
            >
                <Unknown
                    className="edit-icon"
                />
            </div>
            <button
                className="destroy"
                onClick={[Function]}
            />
        </div>
    </div>
</li>,
  "nodes": Array [
    <li
      className="message-wrapper "
>
      <div>
            <div>
                  <img
                        className="img-circle"
                        src="some-url"
                  />
                  <div
                        className="speech-bubble"
                  >
                        <div
                              className="main-message"
                        >
                              some-text
                        </div>
                  </div>
                  <div
                        className="time-message"
                  >
                        22-7-2017 20:05
                  </div>
                  <div
                        className="rename"
                        onClick={[Function]}
                  >
                        <Unknown
                              className="edit-icon"
                        />
                  </div>
                  <button
                        className="destroy"
                        onClick={[Function]}
                  />
            </div>
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Message
        message={
                Object {
                        "createdAt": "2017-07-22T18:05:26Z",
                        "text": "some-text",
                      }
        }
        viewer={
                Object {
                        "avatar": "some-url",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Message {
        "_onDestroyClick": [Function],
        "_onEditClick": [Function],
        "_onInputCancel": [Function],
        "_onInputSave": [Function],
        "_reactInternalInstance": [Circular],
        "_setEditMode": [Function],
        "context": Object {},
        "props": Object {
          "message": Object {
            "createdAt": "2017-07-22T18:05:26Z",
            "text": "some-text",
          },
          "viewer": Object {
            "avatar": "some-url",
          },
        },
        "refs": Object {},
        "state": Object {
          "isEditing": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li
          className="message-wrapper "
>
          <div>
                    <div>
                              <img
                                        className="img-circle"
                                        src="some-url"
                              />
                              <div
                                        className="speech-bubble"
                              >
                                        <div
                                                  className="main-message"
                                        >
                                                  some-text
                                        </div>
                              </div>
                              <div
                                        className="time-message"
                              >
                                        22-7-2017 20:05
                              </div>
                              <div
                                        className="rename"
                                        onClick={[Function]}
                              >
                                        <Unknown
                                                  className="edit-icon"
                                        />
                              </div>
                              <button
                                        className="destroy"
                                        onClick={[Function]}
                              />
                    </div>
          </div>
</li>,
        "_debugID": 2,
        "_renderedOutput": <li
          className="message-wrapper "
>
          <div>
                    <div>
                              <img
                                        className="img-circle"
                                        src="some-url"
                              />
                              <div
                                        className="speech-bubble"
                              >
                                        <div
                                                  className="main-message"
                                        >
                                                  some-text
                                        </div>
                              </div>
                              <div
                                        className="time-message"
                              >
                                        22-7-2017 20:05
                              </div>
                              <div
                                        className="rename"
                                        onClick={[Function]}
                              >
                                        <Unknown
                                                  className="edit-icon"
                                        />
                              </div>
                              <button
                                        className="destroy"
                                        onClick={[Function]}
                              />
                    </div>
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Message
    message={
        Object {
            "createdAt": "2017-07-22T18:05:26Z",
            "text": "some-text",
          }
    }
    viewer={
        Object {
            "avatar": "some-url",
          }
    }
/>,
}
`;
